### DISCUSSION TITLE
How to send custom instructions via instruction data

### DISCUSSION THREAD

**Discussion by sh4hri4roo** [bmresearch/Solnet#381] [Q&A] [answered]
Hello,
how can I send custom instructions via instruction data to my Solana program? Is there any example or step by step guide?

**Comment by hoakbuilds** [comment, marked as answer]
The same way you would interact with any other program. We have a lot of implemented programs, you will need to implement methods like the ones seen [here](https://github.com/bmresearch/Solnet/blob/master/src/Solnet.Programs/SystemProgram.cs#L40-L54) and add those `TransactionInstruction` to the `TransactionBuilder`. 

If you're using anchor for your program, you can use [Solnet.Anchor](https://github.com/bmresearch/Solnet.Anchor) to generate a client to fetch the accounts and create the instructions + submit them.

**Reply by sh4hri4roo** [reply]
Thank you for fast response.
Actually I've done it once with ts and below is my workflow:

`const lamports = 1_000_000_000;`
`const data = Buffer.from(Uint8Array.of(1, ...new BN(lamports).toArray("le", 8)));`
`const initMintIx = new web3.TransactionInstruction({`
`keys: [{...}],`
`programId,`
`data`
`});`
`await connection.sendTransaction(tx, [signer, initial], { skipPreflight: false, preflightCommitment: "confirmed" });`

and it is working perfectly fine but with solnet I get custom error. I think it is because of something wrong I'm doing with data!
Second, I didn't find any samples on how to define data layout for my account like this:

`const uint64 = (property = "uint64") => {  return BufferLayout.blob(8, property); };`
`const publicKey = (property = "publicKey") => { return BufferLayout.blob(32, property); };`
`const MINT_ACCOUNT_DATA_LAYOUT = BufferLayout.struct([BufferLayout.u8("isInitialized") as any,` `publicKey("minter"),uint64("next_num")]};`

and then span it in the createAccount method.

**Reply by hoakbuilds** [reply]
We do not have a similar way to define the data layout, we have always hard coded the size of an account and then implmented methods to decode the data which use layouts like [these](https://github.com/bmresearch/Solnet.Serum/blob/master/Solnet.Serum/Models/OpenOrdersAccount.cs#L20-L91).

**Reply by sh4hri4roo** [reply]
Thanks a lot for the time and fast responses..
